### DISCUSSION TITLE
Version skew feature

### DISCUSSION THREAD

**Discussion by rmarscher** [wakujs/waku#1037] [Ideas]
When new versions of client code are deployed, it is helpful if the framework can help load the new version into open browser windows.

I see some people call this issue `version skew` - https://www.industrialempathy.com/posts/version-skew/. See https://nextjs.org/docs/app/building-your-application/deploying#version-skew and https://svelte.dev/docs/kit/configuration#version

**Comment by dai-shi** [comment]
Yeah, that would be great. I wonder where I would implement it and how. Also, maybe we should learn more about implementation approaches from others.
In the meantime, I guess there should be a user-land solution. I could try something after v0.22 if anybody else has done.

**Reply by rmarscher** [reply]
Perhaps we can use a vite plugin to generate some type of version ID during build and place that in a public static file - like `_version`. We could allow that file name to be changed in the config.

A user-land client solution could then read and poll for that file and do something to load new module versions.

The SvelteKit solution is combined with their framework store that polls for updates.
https://svelte.dev/docs/kit/$app-stores#updated

A similar client-side solution with jotai might be nice.

**Reply by rmarscher** [reply]
Or for a more integrated approach, maybe the client solution should be part of the waku router.

**Reply by dai-shi** [reply]
It depends on how the user-land solution is user friendly. I think for the first challenge, it can be a proof of concept.

The server side logic can be in custom `entries.tsx`, but for the client side, it depends. If we treat it as an "error", an error boundary can catch it and force reload. If it's 200 response.

We need some ways to trap the response. 
https://github.com/dai-shi/waku/blob/1d19f7fc27e38f66616310502e2f96818035668a/packages/waku/src/minimal/client.ts#L194 does it for the minimal client, but we don't have a capability in waku/router/client. 🤔 

The ID generation could be skipped using an environment variable (can be done with a user-land vite plugin too.)

Do we want a periodical poll? I thought it could be only the check with RSC request.

**Reply by rmarscher** [reply]
If we can avoid polling, that would be great. Would we automatically append a client version param in the query string for RSC  requests and then the server would return an error if the version doesn't match - which then the client could handle? There are probably certain applications where the developer would want to control what happens - automatic reload or prompt with a manual reload.

**Reply by dai-shi** [reply]
Yeah, like a http header.

**Reply by jasikpark** [reply]
For our Vite SPA React app we host `/v` as a static file containing the github run ID, and convert the links to fully reload the document on click instead of being SPA links, ensuring their next navigation will reload the page. Seems to work well enough. Polled with react query on a 10m cadence.

**Reply by dai-shi** [reply]
Sounds interesting. While I still don't prefer polling, there might be something I could learn especially around vite config. Do you have anything you can share?

**Reply by jasikpark** [reply]
We actually just have a github actions step that sets it up:

```
    # Prepare static files for building
      - name: Inject unique build id into version file
        run: echo ${{ github.run_id }} > public/v
```

and basically if the app ever sees the value of `v` change while it has been running, that means that the site was redeployed while it was running

**Comment by rmarscher** [comment]
Here is some code that I found referenced while browsing issues in other projects. It attempts to recover from an import error in the browser:
https://gist.github.com/mberneti/28769391cf27f7580a55dedab342c63a

- https://github.com/remix-run/remix/discussions/3016
- https://github.com/sveltejs/kit/issues/5208

**Reply by rmarscher** [reply]
I wonder how this problem is different when using React.lazy vs pre-loading links and navigating to other pages.

**Comment by rmarscher** [comment]
I just found this - https://vite.dev/guide/build#load-error-handling
```
window.addEventListener('vite:preloadError', (event) => {
  window.location.reload() // for example, refresh the page
})
```

> When a new deployment occurs, the hosting service may delete the assets from previous deployments. As a result, a user who visited your site before the new deployment might encounter an import error. This error happens because the assets running on that user's device are outdated and it tries to import the corresponding old chunk, which is deleted. This event is useful for addressing this situation.

**Comment by rmarscher** [comment]
I'm testing a vite plugin to generate a build ID. I'd rather store it somewhere that can be imported by the server. platformData? Or something else? I also don't know if there's a better way to generate the build ID but it seems to be effective with the base template. https://github.com/rmarscher/waku-version-skew-example/blob/main/waku.config.ts#L44

Once we have a build ID, we could pass it down to React somehow and set it in a request header from the Waku client and then compare it on the server. If it doesn't match, the server can return some type of error code that the client can use to reload the modules - or reload the page.

**Reply by rmarscher** [reply]
In the meantime, I can create a client solution that polls `/assets/client-build-id.txt` but not as ideal.

**Reply by dai-shi** [reply]
How about setting the build ID in the http header using Waku middleware and compare it on the client? (Ah, I need to expose enhanceFetch from Router.)

Ideally, I would like a solution that works for static sites, so the http header isn't an option in that case.

**Reply by rmarscher** [reply]
I think I was able to make the build ID available to static components using a Vite virtual module. https://github.com/rmarscher/waku-version-skew-example/commit/e6c5007b4b5a09ae87899a69f1e1eaf81185b00b

Demo here https://waku-version-skew.martiandev.workers.dev/

**Reply by rmarscher** [reply]
I learned that Cloudflare has a header `Cloudflare-Workers-Version-Key`

https://developers.cloudflare.com/workers/configuration/versions-and-deployments/gradual-deployments/#version-affinity

I also found documentation for Vercel's `x-deployment-id` header: https://vercel.com/docs/skew-protection#supported-frameworks

`enhanceFetch` sounds good.

**Reply by dai-shi** [reply]
#1307

**Comment by dai-shi** [comment]
#2072
